### PR TITLE
IDFilter clean-up and extension

### PR DESCRIPTION
**Pull request by hendrikweisser**
Refactoring (at this point largely a rewrite) of the IDFilter library class, and - to a lesser extent - the IDFilter TOPP tool.

Solves #1706.

The library class now uses a more "functional" design based on two aspects: Predicates, that query some aspect of a peptide/protein hit or ID; implemented as structs that define a `bool operator()(...) const`. And two higher-order functions that filter a container based on a predicate, "keepMatchingItems" and "removeMatchingItems".
The filter functions have been given more meaningful names and have been changed to operate on vectors of peptide or protein IDs (which makes them easier to apply and avoids computational overhead). Filters now modify their inputs in-place to avoid unnecessary copying.
Also added some new filter functions for new functionality in the TOPP tool (see below).

Implementation of the IDFilter TOPP tool has been greatly simplified. There are some changes to the parameters:
- "length" and "charge" now accept ranges
- added "whitelist" versions for existing "blacklist" options, and vice versa
- added "whitelist"/"blacklist" filtering by modification
- made the RTPredict filtering options advanced (since nobody uses them)

This also contains changes that solve some const'ness problems in MetaInfoRegistry (see https://github.com/OpenMS/OpenMS/issues/1759#issuecomment-166856061 and https://github.com/OpenMS/OpenMS/commit/edb1546d77de22a76e49002f980497df0cb36ae1).

Updated documentation, tests, and even pyOpenMS!

A remaining problem is that the Doxygen documentation for some of the structs in the IDFilter class doesn't seem to show up.

### CODE REVIEW

**Comment by hendrikweisser** [comment]
One thing I forgot:
In the TOPP tool, "whitelist:proteins" now matches by protein accession only, no longer by sequence. To get matching by sequence, just apply PeptideIndexer before IDFilter, which is much faster and offers relevant options.

File: src/openms/include/OpenMS/FILTERING/ID/IDFilter.h [hunk 1]
```diff
@@ -68,476 +77,813 @@ namespace OpenMS
...
+      {}
+
+      bool operator()(const PeptideHit& hit) const
+      {
+        std::set<String> present_accessions = hit.extractProteinAccessions();
+        for (std::set<String>::iterator it = present_accessions.begin();
+             it != present_accessions.end(); ++it)
         {
-          filtered_hits.push_back(*it);
+          if (accessions.count(*it) > 0) return true;
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FILTERING/ID/IDFilter.h]: is there a specific reason for counting? find is faster

File: src/openms/include/OpenMS/FILTERING/ID/IDFilter.h [hunk 2]
```diff
@@ -68,476 +77,813 @@ namespace OpenMS
...
+      {
+        return (accessions.count(hit.getAccession()) > 0);
       }
 
-      if (!filtered_hits.empty())
+      bool operator()(const PeptideEvidence& evidence) const
       {
-        filtered_identification.setHits(filtered_hits);
-        filtered_identification.assignRanks();
+        return (accessions.count(evidence.getProteinAccession()) > 0);
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FILTERING/ID/IDFilter.h]: see above

File: src/openms/source/FILTERING/ID/IDFilter.cpp [hunk 3]
```diff
@@ -50,622 +52,577 @@ namespace OpenMS
...
+      struct HasMatchingAccession<PeptideEvidence> acc_filter(accessions);
+      // check protein accessions of each peptide hit
+      for (vector<PeptideHit>::iterator hit_it = pep_it->getHits().begin();
+           hit_it != pep_it->getHits().end(); ++hit_it)
       {
-        if (protein_sequences.find(identification.getHits()[i].getSequence().toUnmodifiedString()) != String::npos)
-        {
-          filtered_peptide_hits.push_back(identification.getHits()[i]);
-        }
+        // no non-const "PeptideHit::getPeptideEvidences" implemented, so we
```
Comment (timosachsenberg) [inline src/openms/source/FILTERING/ID/IDFilter.cpp]: ok. maybe we can add it

File: src/openms/source/FILTERING/ID/IDFilter.cpp [hunk 4]
```diff
@@ -50,622 +52,577 @@ namespace OpenMS
...
+                                            peptides)
+  {
+    for (vector<PeptideIdentification>::iterator pep_it = peptides.begin();
+         pep_it != peptides.end(); ++pep_it)
     {
-      ProteinIdentification::ProteinGroup filtered;
-      filtered.accessions.reserve(group_it->accessions.size());
-      for (vector<String>::const_iterator acc_it = group_it->accessions.begin();
-           acc_it != group_it->accessions.end(); ++acc_it)
+      // there's no "PeptideHit::operator<" defined, so we can't use a set nor
```
Comment (timosachsenberg) [inline src/openms/source/FILTERING/ID/IDFilter.cpp]: ok.

File: src/openms/source/METADATA/MetaInfo.cpp [hunk 5]
```diff
@@ -113,27 +113,17 @@ namespace OpenMS
 
   bool MetaInfo::exists(const String & name) const
   {
-    try
+    UInt index = registry_.getIndex(name);
+    if (index != UInt(-1))
```
Comment (timosachsenberg) [inline src/openms/source/METADATA/MetaInfo.cpp]: cast needed?

File: src/tests/topp/IDFilter_5b_output.idXML [hunk 6]
```diff
@@ -23,6 +21,8 @@
 			</PeptideHit>
 			<PeptideHit score="11.38" sequence="FINFGVNVEVLSRFQTK" charge="2" >
 			</PeptideHit>
+			<PeptideHit score="11.37" sequence="UFF" charge="2" protein_refs="PH_0" >
```
Comment (timosachsenberg) [inline src/tests/topp/IDFilter_5b_output.idXML]: why is the order changed?

File: src/topp/IDFilter.cpp [hunk 7]
```diff
@@ -180,485 +145,408 @@ class TOPPIDFilter :
...
-    registerInputFile_("whitelist:proteins", "<file>", "", "filename of a FASTA file containing protein sequences.\n"
-                                                           "All peptides that are not a substring of a sequence in this file are removed\n"
-                                                           "All proteins whose accession is not present in this file are removed.", false);
+    registerTOPPSubsection_("whitelist", "Filtering by whitelisting (only peptides/proteins from a given set can pass)");
+    registerInputFile_("whitelist:proteins", "<file>", "", "Filename of a FASTA file containing protein sequences.\n"
+                                                           "All peptides that are not referencing a protein in this file are removed.\n"
+                                                           "All proteins whose accessions are not present in this file are removed.", false);
     setValidFormats_("whitelist:proteins", ListUtils::create<String>("fasta"));
-    registerFlag_("whitelist:by_seq_only", "Match peptides with FASTA file by sequence instead of accession and disable protein filtering.");
```
Comment (timosachsenberg) [inline src/topp/IDFilter.cpp]: @lars20070 FYI sequence based filtering has been removed

File: src/topp/IDFilter.cpp [hunk 8]
```diff
@@ -180,485 +145,408 @@ class TOPPIDFilter :
     registerTOPPSubsection_("precursor", "Filtering by precursor RT or m/z");
     registerStringOption_("precursor:rt", "[min]:[max]", ":", "Retention time range to extract.", false);
     registerStringOption_("precursor:mz", "[min]:[max]", ":", "Mass-to-charge range to extract.", false);
-    registerFlag_("precursor:allow_missing", "When filtering by precursor RT or m/z, keep peptide IDs with missing precursor information ('RT'/'MZ' meta values)?");
```
Comment (timosachsenberg) [inline src/topp/IDFilter.cpp]: has this been used before? should we be concerned about backward compatibility

**Comment by timosachsenberg** [comment]
Great work. Thanks for taking the lead on this.
I really like it but maybe we should discuss the parameter changes in IDFilter as these might invalidate workflows.

**Comment by timosachsenberg** [comment]
there are some minor issues in CDASH issues:
http://cdash.openms.de/testDetails.php?test=29836359&build=160803

File: src/openms/include/OpenMS/FILTERING/ID/IDFilter.h (hunk 1 above)
Comment (hendrikweisser) [inline src/openms/include/OpenMS/FILTERING/ID/IDFilter.h]: It's a set, so "count" and "find" should be equivalent.

File: src/openms/source/METADATA/MetaInfo.cpp (hunk 5 above)
Comment (hendrikweisser) [inline src/openms/source/METADATA/MetaInfo.cpp]: How do you mean that?

Comment (timosachsenberg) [inline src/openms/source/METADATA/MetaInfo.cpp]: wouldn't index != -1 be enough?

File: src/tests/topp/IDFilter_5b_output.idXML (hunk 6 above)
Comment (hendrikweisser) [inline src/tests/topp/IDFilter_5b_output.idXML]: The hits are now sorted by score. This happens in `IDFilter::assignRanks`, which before was only called for some filters (but possibly multiple times), but now always gets called once in the end. I think that makes the result more predictable.

File: src/topp/IDFilter.cpp (hunk 8 above)
Comment (hendrikweisser) [inline src/topp/IDFilter.cpp]: The functionality provided by this flag had been removed already (see https://github.com/OpenMS/OpenMS/blob/develop/src/topp/IDFilter.cpp#L302). I just cleaned up the left-over parameter.

File: src/openms/source/METADATA/MetaInfo.cpp (hunk 5 above)
Comment (hendrikweisser) [inline src/openms/source/METADATA/MetaInfo.cpp]: I think that might give a "comparing signed to unsigned" warning, but I haven't tried it.

File: src/topp/IDFilter.cpp (hunk 7 above)
Comment (hendrikweisser) [inline src/topp/IDFilter.cpp]: Kind of. The sequence-based filtering here was badly implemented and inflexible. Just use PeptideIndexer instead. See my comment here: https://github.com/OpenMS/OpenMS/pull/1778#issuecomment-173608623

**Comment by hendrikweisser** [comment]
I've added the "explicit" declarations that cpplint complained about.

There seems to be a problem with the Travis builds - they failed with this error:
```
CMake Error at /usr/local/Cellar/cmake/3.0.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:136 (message):
  Could NOT find SEQAN (missing: SEQAN_INCLUDE_DIRS) (Required is at least
  version "1.4.0")
```

**Comment by timosachsenberg** [comment]
I restarted the build. May have just been a hickup...

**Comment by hendrikweisser** [comment]
One of the test builds failed some of the OpenSWATH tests and then ran out of time (although it took less time than some of the other builds?): https://travis-ci.org/OpenMS/OpenMS/jobs/104618551

Strangely, I can't find this build on CDash, so I can't look in more detail. It should be on this page, but it's not in the list: http://cdash.openms.de/index.php?project=OpenMS&date=2016-01-25

**Comment by timosachsenberg** [comment]
Hi @hendrikweisser this is most likely because the macos builds are on a different infrastructure (non-docker) and therefor limited by build time. I think this should be fine.

**Comment by hroest** [comment]
yes, this is a current issue. It is nice to Mac OS X builds but they dont help much if they always fail. Maybe we could simply use travis to *build* but not run the tests on OS X ? And then use the nightlies to detect failure ? I dont see how the current system is really helping

**Comment by hroest** [comment]
see #1791

**Comment by hendrikweisser** [comment]
The one OS X build still timed out, but none of the tests in it failed: https://travis-ci.org/OpenMS/OpenMS/jobs/104618551

I think this should be good to merge, unless anyone still wants to discuss anything (IDFilter parameter changes?).

**Pull request merged**